### PR TITLE
[Csl-985] java sdk add support for http configs

### PR DESCRIPTION
**Pull request by stanlp1**

### CODE REVIEW

File: constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java [hunk 1]
```diff
@@ -53,22 +56,78 @@
...
+
+        client = builder.build();
+        clientWithRetry = builder.retryOnConnectionFailure(true).build();
     }
 
+    /**
+     *
+     * @param config the Http client config
+     */
+    public static void configHttpClient(HttpClientConfig config) {
```
Comment (jjl014) [inline constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java:89]: Can we change this to `setHttpClientConfig`?

File: constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java [hunk 2]
```diff
@@ -53,22 +56,78 @@
...
+    private static OkHttpClient clientWithRetry = client.newBuilder()
         .retryOnConnectionFailure(true)
         .build();
 
     /**
      * @param newClient the HTTP client to use by all instances
      */
-    protected static void setClient(OkHttpClient newClient) {
-        client = newClient;
+    public static void setClient(OkHttpClient newClient) {
```
Comment (jjl014) [inline constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java:66]: Can we change this to `setOkHttpClient` and update the different places where it's used?

We should also update the param to specify OkHttpClient (i.e. `newOkHttpClient the OkHttpClient to use by all instances` or something) 🤔

File: constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java [hunk 3]
```diff
@@ -53,22 +56,78 @@
...
+        }
+
+        if (config.getConnectTimeout() != null) {
+            builder.connectTimeout(config.getConnectTimeout(), TimeUnit.MILLISECONDS);
+        }
+
+        if (config.getConnPoolMaxIdleConnections() != null || config.getConnPoolKeepAliveDuration() != null) {
+            Integer maxIdle = config.getConnPoolMaxIdleConnections();
+            Integer keepAlive = config.getConnPoolKeepAliveDuration();
+            ConnectionPool pool = new ConnectionPool(maxIdle == null ? 5 : maxIdle, keepAlive == null ? 5l : maxIdle, TimeUnit.MILLISECONDS);
```
Comment (jjl014) [inline constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java:108]: What is `5l` (5 L) ?

**Review by jjl014 — COMMENTED** [review]
Looking good so far, left some comments. 👍

File: constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java [hunk 4]
```diff
@@ -53,22 +56,78 @@
...
+        }
+
+        client = builder.build();
+        clientWithRetry = builder.retryOnConnectionFailure(true).build();
+    }
     /**
      * @return the HTTP client used by all instances
      */
-    protected static OkHttpClient getClient() {
+    public static OkHttpClient getClient() {
```
Comment (jjl014) [inline constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java:130]: `getHttpClient` ? 🤔

**Comment by rubikzube** [comment]
We can haz tests around setting the client and setting the client configuration>

File: constructorio-client/src/test/java/io/constructor/client/ConstructorIOHttpClientConfigTest.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+        assertEquals(100, ConstructorIO.getHttpClient().readTimeoutMillis());
+        assertEquals(100, ConstructorIO.getHttpClient().writeTimeoutMillis());
+        assertEquals(500, ConstructorIO.getHttpClient().connectTimeoutMillis());
+        assertEquals(30, ConstructorIO.getHttpClient().dispatcher().getMaxRequests());
+        assertEquals(20, ConstructorIO.getHttpClient().dispatcher().getMaxRequestsPerHost());
+    }
+
+    @Test
+    public void nullConfigVariablesDoNotGetConfigured() {
```
Comment (rubikzube) [inline constructorio-client/src/test/java/io/constructor/client/ConstructorIOHttpClientConfigTest.java:81]: Maybe `setHttpClientConfigWithEmptyConfigShouldReturnDefaults`?

File: constructorio-client/src/main/java/io/constructor/client/HttpClientConfig.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
+package io.constructor.client;
+
+public class HttpClientConfig {
+    private Integer readTimeout;
+    private Integer writeTimeout;
+    private Integer connectTimeout;
+    private Integer connPoolMaxIdleConnections;
```
Comment (rubikzube) [inline constructorio-client/src/main/java/io/constructor/client/HttpClientConfig.java:7]: Maybe don't abbreviate `conn` in general

File: constructorio-client/src/main/java/io/constructor/client/HttpClientConfig.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    private Integer connPoolKeepAliveDuration;
+    private Integer dispatcherMaxRequests;
+    private Integer dispatcherMaxRequestsPerHost;
+
+    HttpClientConfig() {
+        this.readTimeout = null;
```
Comment (rubikzube) [inline constructorio-client/src/main/java/io/constructor/client/HttpClientConfig.java:13]: Can we set these to common defaults instead of null?

**Review by rubikzube — CHANGES_REQUESTED** [review]

**Comment by rubikzube** [comment]
@stanlp1 : are the defaults in the config the okhttp defaults?

**Comment by stanlp1** [comment]
> @stanlp1 : are the defaults in the config the okhttp defaults?

All the defaults here should be the okhttp defaults with the exception of readtimeout and writetimeout. I set those to be the times that we use in the default constructor client if they don't config anything. So if they don't change anything here this should be identical to the default constructor settings.

**Review by rubikzube — APPROVED** [review]

**Pull request merged**